classes/local/url.php

```diff
@@ -0,0 +1,239 @@
+                'null' => NULL_ALLOWED,
+                'default' => '2',
+            ),
+        );
+    }
+
+    /**
+     * Create the hashed field before inserting or updating a record
+     * This runs as the persistent object is created.
+     */
+    protected function before_validate() {
+        $url = $this->get('url');
+        $this->set('urlhash', self::hash_url($url));
+    }
+
+    /**
+     * Reset a node to be recrawled
+     *
+     * @param integer $nodeid node id
+     */
+    public static function reset_for_recrawl($nodeid) {
+
+        global $DB;
+
+        if ($node = new url($nodeid)) {
+
+            $time = crawler::get_config()->crawlstart;
+
+            // Mark all nodes that link to this as needing a recrawl.
+            if ($DB->get_dbfamily() == 'mysql') {
+                $DB->execute("UPDATE {tool_crawler_url} u
+                          INNER JOIN {tool_crawler_edge} e ON e.a = u.id
+                                 SET needscrawl = ?,
+                                     lastcrawled = null,
+                                     priority = ?
+                               WHERE e.b = ?", [$time, TOOL_CRAWLER_PRIORITY_HIGH, $nodeid]);
+            } else {
+                $DB->execute("UPDATE {tool_crawler_url} u
+                                 SET needscrawl = ?,
+                                     lastcrawled = null,
+                                     priority = ?
+                                FROM {tool_crawler_edge} e
+                               WHERE e.a = u.id
+                                 AND e.b = ?", [$time, TOOL_CRAWLER_PRIORITY_HIGH, $nodeid]);
+            }
+            // Delete all edges that point to this node.
+            $DB->delete_records('tool_crawler_edge', ['b' => $nodeid]);
+            // Delete the 'to' node as it may be completely wrong.
+            $DB->delete_records('tool_crawler_url', array('id' => $nodeid) );
+        }
+    }
+
+    /**
+     * Many URLs are in the queue now (more will probably be added)
+     *
+     * @return int size of queue
+     */
+    public function get_queue_size() {
+        global $DB;
+
+        return $DB->get_field_sql("
+                SELECT COUNT(*)
+                  FROM {tool_crawler_url}
+                 WHERE lastcrawled IS NULL
+                    OR lastcrawled < needscrawl");
+    }
+
+    /**
+     * How many URLs have been processed off the queue
+     *
+     * @return int size of processes list
+     */
+    public function get_processed() {
+        global $DB;
+
+        return $DB->get_field_sql("
+                SELECT COUNT(*)
+                  FROM {tool_crawler_url}
+                 WHERE lastcrawled >= ?",
+                                  array(crawler::get_config()->crawlstart));
+    }
+
+    /**
+     * Hash a url
+     * @param $url string the url to hash
+     * @return string the hashed url
+     */
+    public static function hash_url($url) {
+        return md5($url);
```

Comment:
given how broken md5 is probably worth swapping to sha1, even though this isn't a crypto issue it can't hurt

Reply:
Switched